src/extension/vectortiles/src/main/java/org/geotools/renderer/lite/VectorMapRenderUtils.java

```diff
@@ -443,4 +445,33 @@ private static boolean isFeatureTypeStyleActive(FeatureType ftype, FeatureTypeSt
                                 || FeatureTypes.isDecendedFrom(
                                         ftype, null, fts.getFeatureTypeName())));
     }
+
+    /**
+     * This function executes transformations on the vector features e.g. point stacker.
+     *
+     * @param layer the layer to process
+     * @param paintArea the area that is painted on
+     * @param mapScale the current map scale
+     * @param schema the schema of the features
+     * @param features the feature collection typically points
+     * @return the transformed feature collection
+     * @throws IOException
+     */
+    public static FeatureCollection<?, ?> transformIfNecessary(
+            Layer layer,
+            Rectangle paintArea,
+            double mapScale,
+            FeatureType schema,
+            FeatureCollection<?, ?> features)
+            throws IOException {
+
+        List<LiteFeatureTypeStyle> featureTypeStyles =
+                VectorMapRenderUtils.getFeatureStyles(layer, paintArea, mapScale, schema);
+        if (!featureTypeStyles.isEmpty() && featureTypeStyles.get(0).transformation != null) {
+            Expression transform = featureTypeStyles.get(0).transformation;
+            return (FeatureCollection<?, ?>) transform.evaluate(features);
```

Comment:
This should have a little more validation, so that we aren't throwing ClassCast exceptions if there is an invalid transform (for example: heatmap) in the style; at the very least some sort of exception handling with an error that indicates the transform was invalid.

Reply:
Added some exception handling